### PR TITLE
Adding bindings APIs.

### PR DESCRIPTION
**Pull request by philliphoff**
Adds the APIs and registration for input/output bindings, and includes sample bindings for Discord (listening and writing to Discord channels).

### CODE REVIEW

File: samples/DiscordBindingSample/DiscordBinding.cs [hunk 1]
```diff
@@ -0,0 +1,172 @@
+using Azure.Storage.Queues;
+using Dapr.PluggableComponents.Components;
+using Dapr.PluggableComponents.Components.Bindings;
+using Discord;
```
Comment (tmacam) [inline samples/DiscordBindingSample/DiscordBinding.cs:4]: Not sure if there is a `checkstyle` equivalent for dotNet but i think some files are missing the copyright header. But then again, I wonder how required they would be on samples.

File: samples/DiscordBindingSample/Program.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+var app = DaprPluggableComponentsApplication.Create();
+
+app.RegisterService(
+    "discord-binding",
+    serviceBuilder =>
+    {
+        // Use this registration method to have a single pub-sub instance for all components.
+        // serviceBuilder.RegisterBindings<DiscordBinding>();
+
+        // This registration method enables a pub-sub instance per component instance.
```
Comment (tmacam) [inline samples/DiscordBindingSample/Program.cs]: pub-sub?

File: samples/DiscordBindingSample/README.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        "content": "Hello world from Dapr!"
+    },
+    "operation": "SendMessage"
+}
+```
+
+>NOTE: The channel ID can be found by [enabling developer mode](https://discord.com/developers/docs/game-sdk/store#application-test-mode) in Discord, right-clicking the channel name, and selecting `Copy ID` from the context menu.
+
+The application can read messages sent to Discord by exposing a `POST` route that matches the binding name (i.e. `discord-binding`), and accepts a JSON body of the form:
+
```
Comment (tmacam) [inline samples/DiscordBindingSample/README.md:68]: Thats a good README

File: src/Dapr.PluggableComponents/Components/MessageDeliveryHandler.cs [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+/// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
+public delegate Task MessageAcknowledgementHandler<TRequest>(TRequest request);
+
+/// <summary>
+/// Represents the handler used to deliver messages to the (Dapr) client.
+/// </summary>
+/// <param name="response">The response (i.e. message) to deliver to the (Dapr) client.</param>
+/// <param name="onAcknowledgement">A handler to receive acknowledgement of the processed message.</param>
+/// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
+public delegate Task MessageDeliveryHandler<TRequest, TResponse>(TResponse response, MessageAcknowledgementHandler<TRequest>? onAcknowledgement = null);
```
Comment (tmacam) [inline src/Dapr.PluggableComponents/Components/MessageDeliveryHandler.cs:29]: Do we plan to use it with something else than `InputBindingReadRequest` and `InputBindingReadResponse`? Wouldn't it be better to just make it specific to those types and, if required, make it generic later?

File: src/Dapr.PluggableComponents/Components/Bindings/InputBindingReadRequest.cs [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+/// <param name="MessageId">Gets the ID of the processed message.</param>
+/// <remarks>
+/// This type is considered a "request" because it is sent to the component by the (Dapr) client.
+/// </remarks>
+public sealed record InputBindingReadRequest(string MessageId)
+{
+    /// <summary>
+    /// Gets the message response data, if any.
+    /// </summary>
+    public ReadOnlyMemory<byte> ResponseData { get; init; } = ReadOnlyMemory<byte>.Empty;
```
Comment (tmacam) [inline src/Dapr.PluggableComponents/Components/Bindings/InputBindingReadRequest.cs:30]: nice

**Review by tmacam — APPROVED** [review]
LGTM.

Just left a comment about making MessageDeliveryHandler a concrete class rather than a generic.

File: src/Dapr.PluggableComponents/Components/MessageDeliveryHandler.cs (hunk 4 above)
Comment (philliphoff) [reply src/Dapr.PluggableComponents/Components/MessageDeliveryHandler.cs:29]: The same pattern and type are used for the pub-sub streaming read method, hence the generic types.  They started as separate sets of types, then consolidated; to be honest, I don't know which I prefer more.

File: samples/DiscordBindingSample/DiscordBinding.cs (hunk 1 above)
Comment (philliphoff) [reply samples/DiscordBindingSample/DiscordBinding.cs:4]: All of the SDK-related source has the copyright/license headers, but I had intentionally omitted them from the sample source (as we didn't do that for other products).  I'll follow up with to see whether that's something we should be doing.

File: samples/DiscordBindingSample/Program.cs (hunk 2 above)
Comment (philliphoff) [reply samples/DiscordBindingSample/Program.cs]: Good catch.  Updated.

**Pull request merged**